text/0021-state-pruning-and-fast-syncing.md

```diff
@@ -0,0 +1,240 @@
+
+### Pruning summary
+
+Both trie-store as well as storage can be trimmed down by garbage collection, with savings in the latter probably being more substantial than the former.
+
+In both cases the problem can reduced to identifying which root hashes or blocks must be kept and which can be discarded. Once this information is available, garbage collection becomes straight forward.
+
+An "archival node" is a node on which garbage collection is either disabled or restricted to non-transitive state information, *i.e.* it keeps all root hashes except those created in the midst of execution.
+
+### Fast-syncing
+
+Fast syncing, while beneficial in its own right, is also of interest from a storage perspective, as being able to recover information quickly from the network or external sources influences our decisions about what information must be kept.
+
+The "fast" part of fast-syncing, as compared to the status quo, is realized from two factors, namely
+
+* **side-loading** data from outside the network through external sources and
+* **transfering** state, namely trie-store state, which is currently not transferrable at all.
+
+### Side-loading
+
+We create a common trait for Blocks, Deploys or any other transferable object that states how it can be serialized and defines a hash function for its serialized representation, as well as a key to separate the hash space. As an example, a block whose hash is `5A4A55A8CA87B584960DCB0673DC1260292991EC6D2B8F214701B5224FB141D2` would be adressable as `blockv1_5A4A55A8CA87B584960DCB0673DC1260292991EC6D2B8F214701B5224FB141D2`. The `blockv1` denotes a complete scheme, that is it also pins the serialization format. While it is not necessary to separate the hash value space to avoid collisions, the prefix allows for multiple representations, should the format of a serialization or an internal data structure change at some point. We refer to data structures that satisfy this condition as *objects* from here on out.
+
+**Note**: The defunct CEP3 proposed hashing the serialized representation of an object instead of the object itself, which makes it a lot easier for external stores to handle these, as integrity can be checked without deserializing first. However, hashing the internal object minimizes the impact on the existing codebase instead.
+
+We can then define a generic mechanism for requesting these objects from other nodes. The true value however is in being able to serialize these to static external storage.
+
+### S3, HTTP and read/write, read-only stores
+
+As objects are unchanging, they can be readily be side-loaded from an HTTP-server that acts as a read-only store. As an example, should a node need to obtain the block `5A4A55A8CA87B584960DCB0673DC1260292991EC6D2B8F214701B5224FB141D2`, it can follow the following steps:
+
+1. Check one or more configured external HTTP stores by calling `GET store/objects/5A4A55A8CA87B584960DCB0673DC1260292991EC6D2B8F214701B5224FB141D2`
```

Comment:
I'm not sure whether we should be favouring our own network as the first call.  I guess it depends upon whether we see our network as the primary storage source for current data with outside storage being a backup, or vice-versa.  I've been used to the former view, but maybe I need to rethink.

Long-term, I'm not clear on whether we see this off-network storage as a permanent solution or if it's a stop-gap until we build an archival system which would make our network self-reliant and crucially, not dependent on centralised solutions.

Reply:
> I'm not sure whether we should be favouring our own network as the first call. I guess it depends upon whether we see our network as the primary storage source for current data with outside storage being a backup, or vice-versa. I've been used to the former view, but maybe I need to rethink.

I've been used to the former view as well.

Reply:
I think this is sort of the wrong perspective - we should be looking at these as authoritative vs optional, instead of primary vs secondary. Luckily we don't have to deal with conflicting values, since everything is hashed, so the "authority" part is just regarding completeness.

My view is as follows: The decentralized network is the ultimate authority on everything, at it should always (somehow) contain everything considered keepable, while other sources are caches. Put differently, a primary source would always be asked first and considered an authority, while a secondary wouldn't. A cache splits these concerns, one will always want to ask a cache first (speedier) and fall back on the actual authoritative source on a miss.

So in short: Ask the off-network storage first, but always ensure that the network can act as a fallback. In the case of a failure of all external off-network storage, things would work a lot slower, but still work (and external stores could be rebuilt from it).